src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,753 @@
+--   history. This operates under the assumption that all chains
+--   are moving along at roughly the same pace.
+--
+--   Returns an allowlist of blockheights for each chain.
+--
+--   There are a couple of reasons that a blockheight might
+--   not be accessible on a chain:
+--     - the height is below the genesisHeight for that chain
+--     - the height has been compacted away
+resolveTargets :: (Logger logger)
+  => logger
+  -> [ChainId]
+  -> FilePath
+  -> BlockHeightTargets
+  -> IO (Map ChainId (Set BlockHeight))
+resolveTargets logger cids pactDir targets = do
+  case targets of
+    LatestAll -> do
+      -- Get the highest common blockheight across all chains.
+      -- This assumes that all chains are moving along close enough
+      -- to one another.
+      maxCommonRef <- newIORef @BlockHeight (BlockHeight maxBound)
+      pooledFor cids $ \cid -> do
+        withChainDb cid logger pactDir $ \(SQLiteEnv db _) -> do
+          top <- getLatestBlockHeight db
+          atomicModifyIORef' maxCommonRef $ \x -> (min x top, ())
+      maxCommon <- readIORef maxCommonRef
+      pure $ Map.fromList $ List.map (, Set.singleton maxCommon) cids
+
+    TargetAll ts -> do
+      -- Return an allowlist of targets for each chain. There are a couple of
+      -- reasons that a blockheight might not be accessible on a chain:
+      --   - the height is below the genesisHeight for that chain
+      --   - the height has been compacted away
+      targetsRef <- newIORef @(Map ChainId (Set BlockHeight)) $ Map.empty
+      pooledFor cids $ \cid -> do
+        withChainDb cid logger pactDir $ \(SQLiteEnv db _) -> do
+          pooledFor ts $ \target -> do
+            earliest <- getEarliestBlockHeight db
+            if target < earliest
+            then do
+              logFunctionText logger Warn
+                $ "BlockHeight " <> sshow target <> " doesn't exist on Chain " <> chainIdToText cid
+            else do
+              ensureBlockHeightExists db target
+              atomicModifyIORef' targetsRef $ \m -> (Map.insertWith Set.union cid (Set.singleton target) m, ())
+
+      readIORef targetsRef
+
+-- | Resolve a single target.
+--
+--   Returns the resolved BlockHeight, along with the Set of ChainIds that
+--   contain that target.
+resolveTarget :: (Logger logger)
+  => logger
+  -> [ChainId]
+  -> FilePath
+  -> TargetBlockHeight
+  -> IO (Set ChainId, BlockHeight)
+resolveTarget logger cids pactDir target = do
+  let targets = case target of
+        Latest -> LatestAll
+        Target t -> TargetAll (Set.singleton t)
+  resolved <- resolveTargets logger cids pactDir targets
+
+  when (Map.size resolved == 0) $ do
+    error "internal logic error in resolveTarget: resolved was empty"
+
+  when (not $ allEqual $ Map.elems resolved) $ do
+    error "internal logic error in resolveTarget: targets are not the same across all chains"
+
+  -- kinda gross.
+  let height = List.head $ Set.toList $ List.head $ Map.elems resolved
```

Comment:
while writing this i wasn't sure what a better way is. i'm not really sure other than using some custom function instead of head. haven't thought about it.

Reply:
note that the Map is guaranteed to be a singleton, and so is the associated set. so this is safe here. but not resilient to refactoring :)